index.html

```diff
@@ -0,0 +1,452 @@
+  service: String
+  protocolVersion: Int = 2
+}
+
+enum cs_Graph {
+  AUTH @cs_link(to: {
+    exampleOrg_gslb: {
+      service: "gaia/auth-gql-prod", protocolVersion: 1
+    }
+  }),
+  MAPS @cs_link(to: {
+    exampleOrg_gslb: { service: "geo/prod" }
+  })
+}
+```
+
+### cs_OutboundLink
+<view-of src=photos.graphql#cs_OutboundLink
+  title='Definition of cs_OutboundLink'
+  class=no-linum></view-of>
+
+Provides information about an **outbound link**—a connection to a subgraph across which operations can be executed.
+
+### cs_OutboundLinkHTTP
+<view-of src=photos.graphql#cs_OutboundLinkHTTP
+  title='Definition of cs_OutboundLinkHTTP'
+  class=no-linum></view-of>
+
+Provides HTTP-specific configuration for outbound links. **TK: Currently, this is just a URL, but this is the place where we could do a lot more: headers, etc. ~ak**
+
+
+## Scalars
+
+`compose`d schemas **must** include definitions of any scalars they use. Producers **should** include the `@specifiedBy` directive, but it is not mandatory.
+
+### `cs_SelectionSet`
+
+<view-of src=photos.graphql#cs_SelectionSet
+  title='Definition of cs_SelectionSet'
+  class=no-linum
+></view-of>
+
+A GraphQL selection set represented with the same syntax as defined in [the GraphQL SDL](https://spec.graphql.org/draft/#sec-Selection-Sets).
+
+Note: Unlike the federation `_FieldSet` scalar, a `cs_SelectionSet` must parse as a GraphQL selection set, so the selections must be surrounded by braces.
+
+All legal selection set features are allowed. That includes aliases, field arguments, directives, and inline fragments. Named fragment spreads are syntactically valid, but meaningless, as there is nowhere to define named fragments; future versions of this spec may change this.
+
+### `cs_URL`
+
+<view-of src=photos.graphql#cs_URL
+  title='Definition of cs_URL'
+  class=no-linum
+></view-of>
+
+A [URL](https://www.w3.org/Addressing/URL/url-spec.html). Used by [`@cs_link`](#cs_link).
+
+### cs_Never
+
+<view-of src=photos.graphql#cs_Never
+  title='Definition of cs_Never'
+  class=no-linum
+></view-of>
+
+The `cs_Never` scalar is used to indicate the type of a field which **cannot be resolved**. This is typically due to a composition error. Producers **should** provide more specific error information with a `@cs_error` directive on the relevant field.
+
+<view-of src=photos.graphql#User.favorite
+  title='Using <code>cs_Never</code> and <code>@cs_error</code> to mark a field error'
+></view-of>
+
+## Directives
+
+`compose`d schema **must** include the definitions of all directives they use. Directives from this spec **must** be defined exactly as they are specified here. (This ensures that `compose`d schemas pass standard validations, even if those validators do not know of this spec.)
+
+### `@cs_key`
+<view-of src=photos.graphql#@cs_key
+  title='Definition of cs_key'
+  class=no-linum
+></view-of>
+
+Define an entity key.
+
+`@cs_key` is applied to a fragment to mark that fragment as an entity key.
```

Comment:
I like the idea of using fragments instead of string arguments to specify key fields. For consistency, it seems we may want to do something similar for `requires` and `provides`.

Reply:
yeah, we're gonna, haven't gotten around to updating this yet

Reply:
I like the idea abstractly. I am somewhat concerned that tooling (Apollo-provided or third-party) may not deal well with documents that mix type system and executable definitions.  At the very least it seems like things like Studio's schema page won't be able to show fragments where you might imagine it would show directives on schema elements — though that particular page probably will want to handle CSDL specially anyway. Is there anything we can do to help us feel more confident that we're not going to be producing documents that other tooling is going to reject/handle poorly?

For example, how do the graphql-js and GraphQL-Java functions that parse a document and create a GraphQLSchema internal representation handle documents with fragments?  I would guess that they might error on them, or if they don't they probably fail to round-trip them cleanly.

Reply:
A glance at the source of `apollo-server`/`graphql-tools`/`graphql-js` (but not an attempt to read them) suggests that `makeExecutableSchema`/`buildASTSchema` won't error if there are fragments in your schema, but they won't represent them in any way in the GraphQLSchema object so any round-tripping via that object will lose the fragments. This seems like a bad property to have. I'd suggest we avoid using fragments in CSDL because of this.